### PR TITLE
Add Dataverse as source & custom connector as sink options

### PR DESCRIPTION
**Pull request by alexandrakoller**
This PR includes:
- A new data source option for flow & logic app: Dataverse
- A new data sink option for flow & logic app: Custom connector (certified & non-certified)
- Enabled non-certified custom connectors as a data source
- Bugfixes in the custom connector module

### CODE REVIEW

File: scripts/modules/MsIndustryLinks/templates/ingest/customconnector/flow_customconnector.json [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+                "x-ms-content-hint": "TEXT"
+              }
+            },
+            "required": ["payload"]
+          }
+        }
+      }
+    },
+    "actions": {
+      "Map_data": {
```
Comment (ongk) [inline scripts/modules/MsIndustryLinks/templates/ingest/customconnector/flow_customconnector.json:36]: Is the output of this used anywhere?

File: scripts/modules/MsIndustryLinks/templates/ingest/New-IngestionWorkflow.md [hunk 2]
```diff
@@ -46,6 +55,17 @@ The directory where the ingestion workflow template will be saved. If it doesn't
 | Type:          | [String](https://learn.microsoft.com/en-us/powershell/scripting/lang-spec/chapter-04?view=powershell-7.3#431-strings) |
 | Default value: | None                                                                                                                  |
 
+### -AuthConfigFile
```
Comment (ongk) [inline scripts/modules/MsIndustryLinks/templates/ingest/New-IngestionWorkflow.md:58]: -AuthConfigFile needs to be added to the Syntax section

File: scripts/modules/MsIndustryLinks/templates/ingest/IngestionWorkflow.psm1 [hunk 3]
```diff
@@ -92,6 +94,120 @@ function New-LogicAppIngestionWorkflow {
...
+        $parameters = $DataSinkConfig.parameters
+        if ($null -eq $parameters -or 0 -eq @($parameters.psobject.Properties).Count) {
+            throw "No parameters were specified for the custom connector data sink."
+        }
+
+        # Remove the functionality for a custom connector to ingest a single record
+        $definition.actions.PSObject.Properties.Remove("For_each_item")
+
+        # Map the input data to the format of the custom connector API
+        $definition.actions.Map_data.inputs.select = $DataSinkConfig.mapping
```
Comment (ongk) [inline scripts/modules/MsIndustryLinks/templates/ingest/IngestionWorkflow.psm1:171]: Why is this needed for array but not object?

File: scripts/modules/MsIndustryLinks/templates/ingest/customconnector/flow_customconnector.json [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        "type": "Select",
+        "inputs": {
+          "from": "@array(json(triggerBody()['payload']))",
+          "select": {}
+        }
+      },
+      "Ingest_records": {
```
Comment (ongk) [inline scripts/modules/MsIndustryLinks/templates/ingest/customconnector/flow_customconnector.json:43]: This is basically the same definition as "Ingest_record" below. Can you just reuse this and wrap it with a foreach action instead?

File: scripts/modules/MsIndustryLinks/templates/ingest/customconnector/flow_customconnector.json [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        "kind": "Button",
+        "inputs": {
+          "schema": {
+            "type": "object",
+            "properties": {
+              "payload": {
+                "title": "Payload",
+                "type": "string",
+                "x-ms-dynamically-added": true,
+                "description": "The array of items to ingest into Dataverse.",
```
Comment (ongk) [inline scripts/modules/MsIndustryLinks/templates/ingest/customconnector/flow_customconnector.json:26]: Dataverse -> custom connector

**Review by ongk — CHANGES_REQUESTED** [review]

File: scripts/modules/MsIndustryLinks/templates/data_source/DataSourceWorkflow.psm1 [hunk 6]
```diff
@@ -95,8 +95,11 @@ function New-FlowDatasourceWorkflow {
         $template = New-FlowAzureBlobStorageDatasourceWorkflow -WorkflowConfig $WorkflowConfig -WorkflowGuids $WorkflowGuids
         $triggerType = "datasource"
     }
+    elseif ($dataSource -eq "dataverse") {
```
[the remark this answers is not in the document]
Comment (alexandrakoller) [reply scripts/modules/MsIndustryLinks/templates/data_source/DataSourceWorkflow.psm1:98]: This will get updated once I rebase the branch to bring in the new changes from your PR

File: scripts/modules/MsIndustryLinks/templates/data_source/DataSourceWorkflow.psm1 [hunk 7]
```diff
@@ -244,6 +247,45 @@ function New-FlowAzureBlobStorageDatasourceWorkflow {
     return $baseTemplate
 }
 
+function New-FlowDataverseDatasourceWorkflow {
```
[the remark this answers is not in the document]
Comment (alexandrakoller) [reply scripts/modules/MsIndustryLinks/templates/data_source/DataSourceWorkflow.psm1:250]: This PR is for flows only

File: scripts/modules/MsIndustryLinks/templates/ingest/IngestionWorkflow.psm1 [hunk 8]
```diff
@@ -92,6 +94,120 @@ function New-LogicAppIngestionWorkflow {
...
+        $definition.actions.Flow_output.runAfter = @{Ingest_records = @("Succeeded") }
+    }
+    elseif ($inputType -eq "object") {
+        # Remove the functionality for a custom connector to ingest an array of records
+        $definition.actions.PSObject.Properties.Remove("Map_data")
+        $definition.actions.PSObject.Properties.Remove("Ingest_records")
+
+        $definition.actions.For_each_item.actions.Ingest_record.inputs.host.operationId = $DataSinkConfig.connection.operationId
+        $definition.actions.For_each_item.actions.Ingest_record.inputs.host.apiId = $apiId
+        $definition.actions.For_each_item.actions.Ingest_record.inputs.parameters = $DataSinkConfig.mapping
```
[the remark this answers is not in the document]
Comment (alexandrakoller) [reply scripts/modules/MsIndustryLinks/templates/ingest/IngestionWorkflow.psm1:187]: For a custom connector that accepts a single object, you can pass the mapped object directly into the custom connector action. When you have a custom connector that accepts an array, it needs to remap the array using the Select action, where the select action takes the mapping object as input.

File: scripts/modules/MsIndustryLinks/templates/ingest/customconnector/flow_customconnector.json (hunk 1 above)
Comment (alexandrakoller) [reply scripts/modules/MsIndustryLinks/templates/ingest/customconnector/flow_customconnector.json:36]: Yes, the output of Map_data has to be configured in the parameters for custom connectors that accept an array as input. This cant be directly configured in the workflow because the input name of the custom connector can be variable, so hardcoding that into the workflow isn't ideal.

File: scripts/modules/MsIndustryLinks/templates/ingest/IngestionWorkflow.psm1 (hunk 3 above)
Comment (alexandrakoller) [reply scripts/modules/MsIndustryLinks/templates/ingest/IngestionWorkflow.psm1:171]: Is there another more efficient way to remap an array in flows that I have missed?

File: scripts/modules/MsIndustryLinks/templates/ingest/customconnector/flow_customconnector.json (hunk 4 above)
Comment (alexandrakoller) [reply scripts/modules/MsIndustryLinks/templates/ingest/customconnector/flow_customconnector.json:43]: Can do. Was tossing up if this was more ideal than how it is currently. Both require significant configuration. Happy to make the change.

File: scripts/modules/MsIndustryLinks/templates/ingest/IngestionWorkflow.psm1 (hunk 3 above)
Comment (ongk) [reply scripts/modules/MsIndustryLinks/templates/ingest/IngestionWorkflow.psm1]: Sorry, I saw "$DataSinkConfig.mapping" only for the array version. Guess I missed it in the object version.

**Comment by alexandrakoller** [comment]
@ongk I have added logic app workflows and changes to this PR. They are separated into different commits to make reviewing easier.

**Review by ongk — CHANGES_REQUESTED** [review]

File: scripts/modules/MsIndustryLinks/templates/flow_workflow.json.tmpl
Comment (ongk) [inline scripts/modules/MsIndustryLinks/templates/flow_workflow.json.tmpl:1]: Data source type is missing "Dataverse".

File: scripts/modules/MsIndustryLinks/templates/data_source/dataverse/flow_dataverse.json [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+                    "Succeeded"
+                ]
+            },
+            "type": "Workflow",
+            "inputs": {
+                "host": {
+                    "workflowReferenceName": ""
+                },
+                "body": {
+                    "payload": "@string(outputs('Retrieve_data_from_Dataverse')?['body'])"
```
Comment (ongk) [inline scripts/modules/MsIndustryLinks/templates/data_source/dataverse/flow_dataverse.json:60]: I believe this should be: `@string(outputs('Retrieve_data_from_Dataverse')?['body'].value)`

File: scripts/modules/MsIndustryLinks/templates/ingest/IngestionWorkflow.psm1 [hunk 10]
```diff
@@ -78,31 +80,234 @@ function New-LogicAppIngestionWorkflow {
...
+    switch ($dataSinkType) {
+        "customconnector" {
+            $apiName = $DataSinkConfig.properties.name
+            $apiId = Get-LogicAppApiId -DataSourceType $dataSinkType -ApiName $apiName -IsCustomConnectorCertified $DataSinkConfig.isCertified
+
+            Set-LogicAppCustomConnectorDataSinkConfiguration -DataSinkConfig $DataSinkConfig -Definition $definition | Out-Null
+        }
+        "dataverse" {
+            $apiName = Get-ApiName -DataSourceType $dataSinkType
+            $apiId = Get-LogicAppApiId -DataSourceType $dataSinkType -ApiName $apiName
```
Comment (ongk) [inline scripts/modules/MsIndustryLinks/templates/ingest/IngestionWorkflow.psm1:98]: Where is this new `Get-LogicAppApiId` function? I don't see it in the file.

**Review by ongk — CHANGES_REQUESTED** [review]

File: scripts/modules/MsIndustryLinks/templates/ingest/IngestionWorkflow.psm1 (hunk 10 above)
Comment (alexandrakoller) [reply scripts/modules/MsIndustryLinks/templates/ingest/IngestionWorkflow.psm1:98]: Its a shared function from the DataSourceWorkflow module. It will be one that we split out into a module for shared functions and logic that we discussed earlier, along with Get-ApiName.

**Review by ongk — APPROVED** [review]
